### DISCUSSION TITLE
key chains with hydra-like functionality

### DISCUSSION THREAD

**Discussion by mhwombat** [orbitalquark/textadept#225] [Ideas]
I'm working on writing a textadept plug-in like the emacs "hydra" plugin. Similar plugins are also available for neovim and some other editors. While looking into how I would implement it, I realised that the key chain functionality provides most of what I'm looking for. So I wanted to co-ordinate with you to see if you are interested in adding any of this to textadept, or if I should focus developing a separate plugin.

As I see it, the key chain functionality could be fully hydra-like with two added features. The first one is the most important.

1. When a key chain is activated, the status bar (or a pop-up) could display the possible next keys and their actions. Better yet, when defining a key chain, the user could specify some text describing the action that would be displayed instead of the action itself, e.g. "open file" instead of io.open_file".

2. It would also be nice if there was a way to say that an action doesn't end the key chain. For example, suppose I trigger a key chain for "word mode". I would have the right arrow select the next word, the left arrow select the previous word, and "d" delete the word. But I would want to keep the chain active so that I can type "wwwd" to move forward three words and then delete that word. Any other key, or escape, would end the chain.

The advantage of this approach is that the user seldom needs the mouse, and never needs to memorise key sequences. I'd appreciate any thoughts you have on this.

**Comment by orbitalquark** [comment]
I think Textadept could benefit from something like this, but I would have to see a sample implementation of what the end-result might look like. I recommend you develop a plugin and if it looks like core Textadept would benefit from it, I can incorporate it. This has been done in the past like with recordable macros.

**Comment by mhwombat** [comment]
I've written the plugin, check it out! It's the first thing I've written in Lua, so I would appreciate any suggestions you have for improving it.

[textadept-plugin](https://github.com/mhwombat/textadept-hydra)

I'm very impressed with the design of Textadept. It's very lightweight, but powerful enough to do everything I want. I have been searching for an editor like this for years. Now that I've written this plugin, I understand the design of Textadept better, and I doubt there's any real advantage to incorporating any of its functionality it into Textadept. Thank you so much for making this editor available.

**Comment by orbitalquark** [comment]
That looks great! Thanks for sharing :)

**Comment by mhwombat** [comment]
One question: is there a way I can determine how many characters can fit in the status bar? I want to automatically wrap the text as needed.

**Comment by orbitalquark** [comment]
Unfortunately, there is not.

**Comment by snoopy** [comment]
Very nice. I slightly altered the way keybinds are displayed to make things more clear to me: https://files.catbox.moe/at0jnb.webm

**Comment by mhwombat** [comment]
Thank you @snoopy! My next step was to figure out how to do that. Will update the repo shortly.

**Comment by snoopy** [comment]
This seems to trip up some of the textredux replacements such as open, open-recent and the lexer window. I am guessing because of how keybinds are assigned but not sure. Perhaps @rgieseke would be willing to look into this?

**Reply by rgieseke** [reply]
Maybe `textadept-hydra` could handle the `hydra`-mode by switching to a separate `keys.mode`: https://orbitalquark.github.io/textadept/api.html#modes

Right now it's just using custom event handling i think: https://github.com/mhwombat/textadept-hydra/blob/main/init.lua#L173

Textredux uses a key mode per Textredux buffer: https://github.com/rgieseke/textredux/blob/main/core/buffer.lua#L241 and otherwise switches back to Textadept's default keys mode:
https://github.com/rgieseke/textredux/blob/main/core/buffer.lua#L197

Maybe `hydra` just needs to add a switch like this in its `reset_hydra` function?

**Comment by mhwombat** [comment]
I don't think hydra could use key modes? Based on my limited understanding, there's a difference in philosophy in how unexpected/undefined keys are handled.

- Textadept key modes: key that aren't defined in the mode are ignored until the mode is unset. So "unexpected" keys do nothing. This is like VI, I think.
- Hydra: keys that are not defined in the hydra terminate it. So "unexpected" keys do something (and aren't propagated to other event handlers). Maybe this is causing the problem snoopy is seeing. If so, I could add a flag to control this.

I'll install textredux (it was on my todo list) and investigate.

**Reply by rgieseke** [reply]
Unexpected keys in Textredux are handled by the default mode, so it's also possible to go into a hydra mode from Textredux.

**Comment by mhwombat** [comment]
Now that I think about it, I personally always use the escape key to end a hydra, rather than pressing some arbitrary key. So I think I could use key modes, and have a key to end each hydra (which would default to escape).

**Reply by rgieseke** [reply]
So my guessing above was not related to the problem with key bindings above, nonetheless could be useful to be explicit about exiting hydra mode.

**Comment by mhwombat** [comment]
@snoopy I've installed textredux, and I'm not seeing the issue you mentioned. Can you post your init.lua?

**Reply by snoopy** [reply]
None of the dialogs will appear in a textredux list. Snap/quick open is not affected by this, only the normal lists.

```lua
local textredux = require('textredux')
local hydra = require('hydra')

local m_tools = textadept.menu.menubar[_L["Tools"]]
local m_bookmark = m_tools[_L["Bookmarks"]]

local open_hydra = {
  {
    key = 'o',
    help = 'open current dir',
    action = function()
      io.open_file()
    end,
  },
  {
    key = 'r',
    help = 'recent files',
    action = function()
      io.open_recent_file()
    end,
  },
  {
    key = 'l',
    help = 'lexer',
    action = textadept.file_types.select_lexer,
  },
  {
    key = 'm',
    help = 'bookmarks',
    action = function()
      -- m_bookmark[_L["Goto Bookmark..."]][2]
    end,
  },
}

local main_hydra = {
  { key = 'o', help = 'open', action = open_hydra },
}

hydra.keys = {
  { key = 'f10', action = main_hydra },
}

textredux.hijack()

```

**Reply by rgieseke** [reply]
I think the problem is that Textredux patches the specific keys. Hydra defines its own independent key bindings.

So you would have to define them like
```lua
local open_hydra = {
  {
    key = 'o',
    help = 'open current dir',
    action = textredux.fs.open_file,
  },
```

**Comment by mhwombat** [comment]
@snoopy I just updated the textadept-hydra repo. This new version deactivates hydra while a non-hydra action runs, and then re-activates it afterward. So I think that, combined with the change suggested by @rgieseke (e.g., to use `textredux.fs.open_file` instead of `io.open_file()` should have hydra and textredux getting along nicely.

I'm also looking at using key modes instead of event handling.
